ppo/agents/vanilla_ppo.py

```diff
@@ -29,8 +34,48 @@ def __init__(
     def observe_first(self, timestep: dm_env.TimeStep) -> None:
         self.replay_buffer.add_first(timestep)
 
-    def observe(self, value: float, log_probability: float, action: np.ndarray, next_timestep: dm_env.TimeStep) -> None:
-        self.replay_buffer.add(value, log_probability, action, next_timestep)
+    def observe(self, value: float, log_probability: float, action: np.ndarray, next_timestep: dm_env.TimeStep, advantage: float) -> None:
+        self.replay_buffer.add(value, log_probability, action, next_timestep, advantage)
 
     def update(self):
-        pass
\ No newline at end of file
+        pass
+
+    def prob_ratio(
+        self,
+        policy_params: hk.Params,
+        observation: chex.Array, 
+        action: chex.Array, 
+        action_log_probs_old: chex.Array    
+    ) -> chex.Array:
+        mu, sigma = self.policy_network.apply(policy_params, observation)
+        action_log_probs = rlax.gaussian_diagonal().logprob(action, mu, sigma)
+        ratio = jnp.exp( (action_log_probs) - (action_log_probs_old))
+        return ratio
+
+    def normalize_advantage(self, advantage: chex.Array) -> chex.Array:
+        mean = jnp.mean(advantage)
+        std = jnp.std(advantage)
+        normalized_advantage = (advantage - mean)/(std + 1e-8) # add a little jitter to avoid division by 0
+        return normalized_advantage
+    
+    def policy_loss(
+        self,
+        policy_params: hk.Params,
+        observation,
+        action: np.ndarray,
+        advantage: chex.Array,
+        action_log_probs_old: chex.Array,
+        type: str = 'regular'
+    ) -> chex.Array:
+        mu, sigma = self.policy_network.apply(policy_params, observation)
+        action_log_probs = rlax.gaussian_diagonal().logprob(action, mu, sigma)
+
+        if type == 'clipped':
```

Comment:
Have you tried to differentiate it? I am afraid you cannot differentiate a function with an "if" loop inside.